trin-execution/src/execution.rs

```diff
@@ -115,176 +111,79 @@ impl State {
-
-            match self.database.db.get(keccak256(beneficiary))? {
-                Some(account_bytes) => {
-                    let mut account: Account = Decodable::decode(&mut account_bytes.as_slice())?;
-                    account.balance += U256::from(reward);
-                    self.database
-                        .db
-                        .put(keccak256(beneficiary), alloy_rlp::encode(account))?;
-                }
-                None => {
-                    let mut account = Account::default();
-                    account.balance += U256::from(reward);
-                    self.database
-                        .db
-                        .put(keccak256(beneficiary), alloy_rlp::encode(account))?;
-                }
+                .await
+                .get_next_block()
+                .await?
+                .clone();
+            stop_timer(timer);
+
+            execution_context.execute_block(&block)?;
+            last_state_root = block.header.state_root;
+
+            // Commit the bundle if we have reached the limits, to prevent to much memory usage
+            // We won't use this during the dos attack to avoid writing empty accounts to disk
+            if !(2_200_000..2_700_000).contains(&block_number)
+                && should_we_commit_block_execution_early(
+                    block_number - start,
+                    execution_context.bundle_size_hint() as u64,
+                    execution_context.cumulative_gas_used(),
+                    range_start.elapsed(),
+                )
+            {
+                break;
             }
         }
 
-        // check if dao fork, if it is drain accounts and transfer it to beneficiary
-        if block.header.number == get_spec_block_number(SpecId::DAO_FORK) {
-            process_dao_fork(&self.database)?;
-        }
-        stop_timer(timer);
+        let root_with_trie_diff = execution_context.commit_bundle(last_state_root)?;
 
-        let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["get_root_with_trie_diff"]);
-        let RootWithTrieDiff {
-            root,
-            trie_diff: changed_nodes,
-        } = self.get_root_with_trie_diff()?;
-        if root != block.header.state_root {
-            panic!(
-                "State root doesn't match! Irreversible! Block number: {}",
-                self.execution_position.block_execution_number()
-            )
-        }
-        stop_timer(timer);
-
-        let timer = start_timer_vec(
-            &BLOCK_PROCESSING_TIMES,
-            &["increase_block_execution_number"],
-        );
-        self.execution_position
-            .increase_block_execution_number(self.database.db.clone(), root)?;
+        let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["set_block_execution_number"]);
+        self.execution_position.set_next_block_number(
+            self.database.db.clone(),
+            execution_context.block_number() + 1,
```

Comment:
nit: I don't like that we are getting `execution_context.block_number()`, considering that we should what block we executed. We can maybe combine it with `last_state_root`. I think we can do something like this:


```rust
let mut last_executed_header = None;
for block_number in start..=end {
    ...
    execution_context.execute_block(&block)?;
    last_executed_header = block.header;
    ...
}

let Some(last_executed_header) = last_executed_header else {
    // no blocks were executed
    // either return error, since this can happen only if end<start
    // or something like this:
    return Ok(RootWithTrieDiff {
        root: self.execution_position.state_root(),
        changed_nodes: HashMap::new(),
    });
};

let root_with_trie_diff = execution_context.commit_bundle()?;
ensure!(root_with_trie_diff.root == last_executed_header.state_root, "..");

let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["set_block_execution_number"]);
self.execution_position.set_next_block_number(
    self.database.db.clone(),
    last_executed_header.block_number + 1,
    last_executed_header.state_root,
)?; // nit you can even rename set_next_block_number to update and only pass `last_executed_header` as argument
stop_timer(timer);

Ok(root_with_trie_diff)
```

Reply:
If no blocks were executed with the current design we have no way to get the last header.

I added an ensure at the top of the function which returns an error if end is less then start, which start is now `self.execution_position.next_block_number()`

Reply:
So I will just do `last_executed_block_number` because the suggestion isn't possible in certain edge cases as well